lib/imap/sync.rb

```diff
@@ -0,0 +1,223 @@
+        # in Email::Reciever by matching Message-Ids).
+        Rails.logger.warn("[IMAP] UIDVALIDITY does not match, invalidating IMAP cache and resync emails for #{@group.name} (#{@group.id}) / #{@group.imap_mailbox_name}.")
+        @group.imap_last_uid = 0
+      end
+
+      if idle
+        raise 'IMAP IDLE is disabled' if !SiteSetting.enable_imap_idle
+
+        if !@provider.can?('IDLE')
+          Rails.logger.warn("[IMAP] IMAP server for #{@group.name} (#{@group.id}) does not support IDLE.")
+          return
+        end
+
+        last_response_name = nil
+        @provider.imap.idle do |resp|
+          if resp.kind_of?(Net::IMAP::UntaggedResponse) && resp.name == 'EXISTS'
+            last_response_name = resp.name
+            @provider.imap.idle_done
+          end
+        end
+
+        old_uids = []
+        new_uids = @provider.imap.uid_search('NOT SEEN').filter { |uid| uid > @group.imap_last_uid }
+      else
+        # Fetching UIDs of old (already imported into Discourse, but might need
+        # update) and new (not downloaded yet) emails.
+        if @group.imap_last_uid == 0
+          old_uids = []
+          new_uids = @provider.uids
+        else
+          old_uids = @provider.uids(to: @group.imap_last_uid) # 1 .. seen
+          new_uids = @provider.uids(from: @group.imap_last_uid + 1) # seen+1 .. inf
+        end
+      end
+
+      Rails.logger.debug("[IMAP] Remote email server has #{old_uids.size} old emails and #{new_uids.size} new emails.")
+      all_new_count = new_uids.size
+
+      import_mode = @import_limit > -1 && new_uids.size > @import_limit
+      old_uids = old_uids.sample(@old_emails_limit).sort! if @old_emails_limit > 0
+      new_uids = new_uids[0..@new_emails_limit] if @new_emails_limit > 0
+
+      if old_uids.present?
+        Rails.logger.debug("[IMAP] Syncing #{old_uids.size} randomly-selected old emails.")
+        emails = @provider.emails(@group.imap_mailbox_name, old_uids, ['UID', 'FLAGS', 'LABELS'])
+        emails.each do |email|
+          Jobs.enqueue(:sync_imap_email,
+            group_id: @group.id,
+            mailbox_name: @group.imap_mailbox_name,
+            uid_validity: @status[:uid_validity],
+            email: email,
+          )
+        end
+      end
+
+      if new_uids.present?
+        Rails.logger.debug("[IMAP] Syncing #{new_uids.size} new emails (oldest first).")
+        emails = @provider.emails(@group.imap_mailbox_name, new_uids, ['UID', 'FLAGS', 'LABELS', 'RFC822'])
+        emails.each do |email|
+          # Synchronously process emails because the order of emails matter
+          # (for example replies must be processed after the original email
+          # to have a topic where the reply can be posted).
+          begin
+            receiver = Email::Receiver.new(email['RFC822'],
+              allow_auto_generated: true,
+              import_mode: import_mode,
+              destinations: [{ type: :group, obj: @group }],
```

Comment:
`obj` here is fairly generic. I would prefer `group:` or `target:` to be more consistent with the rest of our codebase.

Reply:
I agree that `obj` is generic, but I believe that is how it is supposed to be, because `type` specifies what `obj` is.

I think I can just pass `@group` and perform a type check with `is_a?`, but will have to do a quick benchmark first because I have a feeling it will be a lot slower.